DOCSP-37511 Indexes (#59)
mongodb/docs-pymongo: source/indexes/unique-index.txt

.. _pymongo-unique-index:

==============
Unique Indexes
==============

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: index, query, optimization, efficiency

Overview
--------

Unique indexes ensure that the indexed fields do not store duplicate values. By
default, MongoDB creates a unique index on the ``_id`` field during the creation

> of a collection. To create a unique index, specify the field or combination of
> fields that you want to prevent duplication on and set the ``unique`` option to
> ``True``.

Reviewer: S: Might be easier as a list:

```suggestion
of a collection. To create a unique index, perform the following steps:

- Specify the field or combination of fields that you want to prevent duplication on.
- Set the ``unique`` option to``True``.
```

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``sample_mflix.movies`` collection
from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
free MongoDB Atlas cluster and load the sample datasets, see the
:ref:`<pymongo-get-started>`.

Create a Unique Index
---------------------

> The following example creates a unique, descending index on the ``theaterId`` field:

Reviewer: The next section says "ascending wildcard index", with no comma. I'd probably use the same pattern here:

```suggestion
The following example creates a descending unique index on the ``theaterId`` field:
```

.. literalinclude:: /includes/indexes/indexes.py
   :language: python
   :start-after: start-index-geo-query
   :end-before: end-index-geo-query

.. important::

   If you perform a write operation that stores a duplicate value that violates

>    the unique index, the drier raises a ``DuplicateKeyException``, and MongoDB

Reviewer: ```suggestion
   the unique index, the driver raises a ``DuplicateKeyException``, and MongoDB
```

Author: washer*

   throws an error resembling the following:

   .. code-block:: none
      :copyable: false

>       E11000 duplicate key error index

Reviewer: Move to Troubleshooting section?

Author: Good call
